File: sdk/samples/all in on code/Visual Studio 2008/VBWin7TaskbarJumpList/HelperMethod.vb
'********************************** Module Header **********************************\
' Module Name:  HelperMethods.cs
' Project:      CSWin7TaskbarJumpList
' Copyright (c) Microsoft Corporation.
' 
' The file contains the helper methods tohandle Admin session check, restart 
' application to elevate the user session, register/unregister application ID and 
' file handle, validate file name, and create files under system temp folder.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'***********************************************************************************/

#Region "Imports directive"
Imports System.Security.Principal
Imports System.IO
Imports Microsoft.Win32
#End Region


Public Class HelperMethod

    ' Shared registry key (HKCU or HKCR)
    Shared classRoot As RegistryKey


    '  Helper method to check whether the current application is runas Admin
    Public Shared Function IsAdmin() As Boolean
        Dim id As WindowsIdentity = WindowsIdentity.GetCurrent()
        Dim principal As WindowsPrincipal = New WindowsPrincipal(id)
        Return principal.IsInRole(WindowsBuiltInRole.Administrator)
    End Function


    ' Helper method to restart the current application to evelate it to Admin session
    Public Shared Sub RestartApplicationAsAdmin()
        Dim proc As ProcessStartInfo = New ProcessStartInfo()
        proc.UseShellExecute = True
        ' Get the current application directory
        proc.WorkingDirectory = Environment.CurrentDirectory
        ' Get the current application executable file path
        proc.FileName = Application.ExecutablePath
        proc.Verb = "runas"  ' Elevate the privilege
        Process.Start(proc)  ' Restart the current process
    End Sub


    ' Helper method to check whether the current application has registered some file handle
    Public Shared Function IsApplicationRegistered(ByVal appid As String) _
    As Boolean
        Try
            ' Open the current application's AppID key under HKCR
            Dim progIdKey As RegistryKey = Registry.ClassesRoot.OpenSubKey(appid)

            If Not progIdKey Is Nothing Then
                progIdKey.Close()
                Return True
            End If
        Catch ex As Exception
            MessageBox.Show(ex.Message, ex.Source)
        End Try
        Return False
    End Function


    ' Call InternalRegisterFileAssociations method to register the file handle
    Public Shared Sub RegisterFileAssociations( _
    ByVal progId As String, ByVal registerInHKCU As Boolean, ByVal appId As String, _
    ByVal openWith As String, ByVal ParamArray extensions() As String)
        InternalRegisterFileAssociations(False, progId, registerInHKCU, appId, openWith, extensions)
    End Sub


    ' Call InternalRegisterFileAssociations method to unregister the file handle
    Public Shared Sub UnregisterFileAssociations( _
    ByVal progId As String, ByVal registerInHKCU As Boolean, ByVal appId As String, _
    ByVal openWith As String, ByVal ParamArray extensions() As String)
        InternalRegisterFileAssociations(True, progId, registerInHKCU, appId, openWith, extensions)
    End Sub


    ' Private helper method to register/unregister application file handle
    Private Shared Sub InternalRegisterFileAssociations( _
    ByVal unregister As Boolean, ByVal progId As String, ByVal registerInHKCU As Boolean, _
    ByVal appId As String, ByVal openWith As String, ByVal associationsToRegister() As String)
        Try
            ' Check whether to register the file handle under HKCU or HKCR
            If registerInHKCU Then
                classRoot = Registry.CurrentUser.OpenSubKey("Software\Classes")
            Else
                classRoot = Registry.ClassesRoot
            End If

            ' First of all, unregister the file handle
            For Each assoc In associationsToRegister
                UnregisterFileAssociation(progId, assoc)
            Next
            UnregisterProgId(progId)

            ' Register the application ID and the file handle for each file extension
            If Not unregister Then
                RegisterProgId(progId, appId, openWith)
                For Each assoc In associationsToRegister
                    RegisterFileAssociation(progId, assoc)
                Next
            End If
        Catch ex As Exception
            MessageBox.Show(ex.Message)
        End Try
    End Sub


    ' Register the application ID
    Private Shared Sub RegisterProgId(ByVal progId As String, ByVal appId As String, ByVal openWith As String)
        Dim progIdKey = classRoot.CreateSubKey(progId)
        progIdKey.SetValue("FriendlyTypeName", "@shell32.dll,-8975")
        progIdKey.SetValue("DefaultIcon", "@shell32.dll,-47")
        progIdKey.SetValue("CurVer", progId)
        progIdKey.SetValue("AppUserModelID", appId)
        Dim shell = progIdKey.CreateSubKey("shell")
        shell.SetValue(String.Empty, "Open")
        shell = shell.CreateSubKey("Open")
        shell = shell.CreateSubKey("Command")
        shell.SetValue(String.Empty, openWith)
        shell.Close()
        progIdKey.Close()
    End Sub


    ' Unregister the application ID
    Private Shared Sub UnregisterProgId(ByVal progId As String)
        classRoot.DeleteSubKeyTree(progId)
    End Sub


    ' Register the file handle
    Private Shared Sub RegisterFileAssociation(ByVal progId As String, ByVal extension As String)
        Dim openWithKey = classRoot.CreateSubKey(Path.Combine(extension, "OpenWithProgIds"))
        openWithKey.SetValue(progId, String.Empty)
        openWithKey.Close()
    End Sub


    ' Unregister the file handle
    Private Shared Sub UnregisterFileAssociation(ByVal progId As String, ByVal extension As String)
        Dim openWithKey = classRoot.CreateSubKey(Path.Combine(extension, "OpenWithProgIds"))
        openWithKey.DeleteValue(progId)
        openWithKey.Close()
    End Sub


    ' A helper method to check whether a file name is valid
    Public Shared Function CheckFileName(ByVal fileName As String) As Boolean
        If Not fileName.IndexOfAny(Path.GetInvalidFileNameChars()) = -1 Then
            MessageBox.Show("Please use only characters that are allowed in file names.")
            Return False
        End If
        Return True
    End Function


    ' A helper method to create a file in the temp folder
    Public Shared Function GetTempFileName(ByVal fileName As String)
        Dim path As String = System.IO.Path.Combine(System.IO.Path.GetTempPath(), fileName + ".txt")
        ' Ensure the file exists
        File.Create(path).Close()
        Return path
    End Function

End Class

File: sdk/samples/all in on code/Visual Studio 2008/CSWin7TaskbarJumpList/HelperMethod.cs
/************************************** Module Header *************************************\
* Module Name:  HelperMethods.cs
* Project:      CSWin7TaskbarJumpList
* Copyright (c) Microsoft Corporation.
* 
* The file contains the helper methods tohandle Admin session check, restart application to 
* elevate the user session, register/unregister application ID and file handle, validate 
* file name, and create files under system temp folder.
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\******************************************************************************************/

#region Using directive
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Win32;
using System.Windows.Forms;
using System.IO;
using System.Security.Principal;
using System.Diagnostics;
#endregion


namespace CSWin7TaskbarJumpList
{
    public static class HelperMethod
    {
        // Static registry key (HKCU or HKCR)
        static RegistryKey classesRoot;

        // Helper method to check whether the current application is runas Admin
        public static bool IsAdmin()
        {
            WindowsIdentity id = WindowsIdentity.GetCurrent();
            WindowsPrincipal principal = new WindowsPrincipal(id);
            return principal.IsInRole(WindowsBuiltInRole.Administrator);
        }

        // Helper method to restart the current application to evelate it
        public static void RestartApplicationAsAdmin()
        {
            ProcessStartInfo proc = new ProcessStartInfo();
            proc.UseShellExecute = true;
            // Get the current application directory
            proc.WorkingDirectory = Environment.CurrentDirectory;
            // Get the current application executable file path
            proc.FileName = Application.ExecutablePath;
            proc.Verb = "runas";  // Elevate the privilege
            Process.Start(proc);  // Restart the current process
        }

        // Helper method to check whether the application has registered some file handle
        public static bool IsApplicationRegistered(string appId)
        {
            try
            {
                // Open the current application's AppID key under HKCR
                using (RegistryKey progIdKey = Registry.ClassesRoot.OpenSubKey(appId))
                {
                    return progIdKey != null;
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, ex.Source);
            }
            return false;
        }

        // Call InternalRegisterFileAssociations to register the file handle
        public static void RegisterFileAssociations(string progId, bool registerInHKCU, 
            string appId, string openWith, params string[] extensions)
        {
            InternalRegisterFileAssociations(false, progId, registerInHKCU, appId, openWith, extensions);
        }

        // Call InternalRegisterFileAssociations method to unregister the file handle
        public static void UnregisterFileAssociations(string progId, bool registerInHKCU,
            string appId, string openWith, params string[] extensions)
        {
            InternalRegisterFileAssociations(true, progId, registerInHKCU, appId, openWith, extensions);
        }

        // Private helper method to register/unregister application file handle
        private static void InternalRegisterFileAssociations(bool unregister, string progId,
            bool registerInHKCU, string appId, string openWith, string[] associationsToRegister)
        {
            try
            {
                // Check whether to register the file handle under HKCU or HKCR
                if (registerInHKCU)
                    classesRoot = Registry.CurrentUser.OpenSubKey(@"Software\Classes");
                else
                    classesRoot = Registry.ClassesRoot;

                // First of all, unregister the file handle
                Array.ForEach(associationsToRegister,
                    assoc => UnregisterFileAssociation(progId, assoc));
                UnregisterProgId(progId);

                // Register the application ID and the file handle for each file extension
                if (!unregister)
                {
                    RegisterProgId(progId, appId, openWith);
                    Array.ForEach(associationsToRegister, 
                        assoc => RegisterFileAssociation(progId, assoc));
                }
            }
            catch (Exception e)
            {
                MessageBox.Show(e.Message);
            }
        }

        // Register the application ID
        private static void RegisterProgId(string progId, string appId, string openWith)
        {
            RegistryKey progIdKey = classesRoot.CreateSubKey(progId);
            progIdKey.SetValue("FriendlyTypeName", "@shell32.dll,-8975");
            progIdKey.SetValue("DefaultIcon", "@shell32.dll,-47");
            progIdKey.SetValue("CurVer", progId);
            progIdKey.SetValue("AppUserModelID", appId);
            RegistryKey shell = progIdKey.CreateSubKey("shell");
            shell.SetValue(String.Empty, "Open");
            shell = shell.CreateSubKey("Open");
            shell = shell.CreateSubKey("Command");
            shell.SetValue(String.Empty, openWith);
            shell.Close();
            progIdKey.Close();
        }

        // Unregister the application ID
        private static void UnregisterProgId(string progId)
        {
            classesRoot.DeleteSubKeyTree(progId);
        }

        // Register the file handle
        private static void RegisterFileAssociation(string progId, string extension)
        {
            RegistryKey openWithKey = classesRoot.CreateSubKey(Path.Combine(extension, "OpenWithProgIds"));
            openWithKey.SetValue(progId, String.Empty);
            openWithKey.Close();
        }

        // Unregister the file handle
        private static void UnregisterFileAssociation(string progId,
            string extension)
        {
            RegistryKey openWithKey = classesRoot.CreateSubKey(Path.Combine(extension, "OpenWithProgIds"));
            openWithKey.DeleteValue(progId);
            openWithKey.Close();
        }

        // A helper method to check whether a file name is valid
        public static bool CheckFileName(string fileName)
        {
            if (fileName.IndexOfAny(Path.GetInvalidFileNameChars()) != -1)
            {
                MessageBox.Show("Please use only characters that are allowed in file names.");
                return false;
            }
            return true;
        }

        // A helper method to create a file in the temp folder
        public static string GetTempFileName(string fileName)
        {
            string path = Path.Combine(System.IO.Path.GetTempPath(), fileName + ".txt");
            File.Create(path).Close();  // Ensure the file exists
            return path;
        }
    }
}
